File: problems/medium/2. add_two_numbers.php
<?php
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */
class ListNode {
    public $val = 0;
    public $next = null;
    function __construct($val) { $this->val = $val; }
}

class Solution {

    /**
     * @param ListNode $l1
     * @param ListNode $l2
     * @return ListNode
     */
    function addTwoNumbers($l1, $l2) {
        $carry = 0;
        $header = new ListNode(0);
        $pointer = $header;

        while($l1 != null && $l2 != null) {
            $v1 = $l1->val;
            $v2 = $l2->val;
            $sum = $v1 + $v2 + $carry;
            if ($sum > 9) {
                $sum = $sum % 10;
                $carry = 1;
            } else {
                $carry = 0;
            }
            $pointer->next = new ListNode($sum);
            $pointer = $pointer->next;
            $l1 = $l1->next;
            $l2 = $l2->next;
        }

        while($l1 != null) {
            $v1 = $l1->val;
            $sum = $v1 + $carry;
            if ($sum > 9) {
                $sum = $sum % 10;
                $carry = 1;
            } else {
                $carry = 0;
            }
            $pointer->next = new ListNode($sum);
            $pointer = $pointer->next;
            $l1 = $l1->next;
        }

        while($l2 != null) {
            $v2 = $l2->val;
            $sum = $v2 + $carry;
            if ($sum > 9) {
                $sum = $sum % 10;
                $carry = 1;
            } else {
                $carry = 0;
            }
            $pointer->next = new ListNode($sum);
            $pointer = $pointer->next;
            $l2 = $l2->next;
        }

        if ($carry != 0) {
            $pointer->next = new ListNode($carry);
        }

        return $header->next;
    }
}

File: codetop/top100/2.add_two_numbers.go
package top100

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummyHead := &ListNode{}
	prev := dummyHead
	carry := 0
	for l1 != nil && l2 != nil {
		prev.Next = &ListNode{}
		sum := l1.Val + l2.Val + carry
		prev.Next.Val = sum % 10
		carry = sum / 10
		prev, l1, l2 = prev.Next, l1.Next, l2.Next
	}
	for l1 != nil {
		prev.Next = &ListNode{}
		sum := l1.Val + carry
		prev.Next.Val = sum % 10
		carry = sum / 10
		prev, l1 = prev.Next, l1.Next
	}
	for l2 != nil {
		prev.Next = &ListNode{}
		sum := l2.Val + carry
		prev.Next.Val = sum % 10
		carry = sum / 10
		prev, l2 = prev.Next, l2.Next
	}
	if carry > 0 {
		prev.Next = &ListNode{Val: carry}
	}
	return dummyHead.Next
}
